src/mod_zone_difficulty_scripts.cpp

```diff
@@ -537,6 +537,72 @@ class mod_zone_difficulty_worldscript : public WorldScript
         sZoneDifficulty->IsDebugInfoEnabled = sConfigMgr->GetOption<bool>("ModZoneDifficulty.DebugInfo", false);
         sZoneDifficulty->LoadMapDifficultySettings();
     }
+
+    void OnAfterUpdateEncounterState(Map* map, EncounterCreditType type, uint32 creditEntry, Unit* source, Difficulty difficulty_fixed, DungeonEncounterList const* encounters, uint32 dungeonCompleted, bool updated)
+    {
+        LOG_ERROR("sql.sql", "Encounter completed");
+        // todo: Apparently this doesn't fire.
+    }
+};
+
+class mod_zone_difficulty_dungeonmaster : public CreatureScript
+{
+public:
+    mod_zone_difficulty_dungeonmaster() : CreatureScript("mod_zone_difficulty_dungeonmaster") { }
+
+    bool OnGossipSelect(Player* player, Creature* /*creature*/, uint32 /*sender*/, uint32 action) override
+    {
+        if (action == 100)
+        {
+            LOG_ERROR("sql.sql", "Try turn on");
+            bool CanTurnOn = true;
+            // Was a single encounter completed on normal mode?
+            if (sZoneDifficulty->HardmodeInstanceData[player->GetMap()->GetInstanceId()].CompletedEncounterOnNormal != true)
+            {
+                CanTurnOn = false;
+                LOG_ERROR("sql.sql", "CompletedEncounterOnNormal");
+            }
+
+            if (!player->GetInstanceScript() || !player->GetInstanceScript()->IsEncounterInProgress())
+            {
+                CanTurnOn = false;
+                LOG_ERROR("sql.sql", "IsEncounterInProgress");
+            }
+
+            if (CanTurnOn == true)
+            {
+                sZoneDifficulty->HardmodeInstanceData[player->GetMap()->GetInstanceId()].HardmodeOn = true;
+                LOG_ERROR("sql.sql", "Turn on");
+            }
+
+            CloseGossipMenuFor(player);
+        }
+        else if (action == 101)
+        {
+            sZoneDifficulty->HardmodeInstanceData[player->GetMap()->GetInstanceId()].HardmodeOn = false;
```

Comment:
same as above, store `sZoneDifficulty->HardmodeInstanceData[player->GetMap()->GetInstanceId()]` inside `someVariable` and then do `if (someVariable) { someVariable.HardmodeOn = false }` 

`else` print some error thatt the data could not be found